C7Engine/AI/Pathing/DijkstrasLandAlgorithm.cs

```diff
@@ -0,0 +1,87 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using C7GameData;
+
+namespace C7Engine.Pathing
+{
+	/**
+	 * Uses Dijkstra's Algorithm to find a path between two tiles.
+	 * Advantages: Finds the shortest path accounting for tile movement cost
+	 * Disadvantages: Expensive, will exhaustively search every tile on the continent
+	 *
+	 * Notes: - Edge weight is defined by the movement cost of the 2nd node:
+	 *          w(u, v) = movement_cost(v)
+	 */
+	public class DijkstrasLandAlgorithm : PathingAlgorithm
+	{
+		//N.B. This should really be static, but we can't put a static method on interfaces, so it isn't.
+		public TilePath PathFrom(Tile start, Tile destination)
+		{
+			// shortest distance from start to each tile on the continent
+			Dictionary<Tile, int> dist = new Dictionary<Tile, int>();
+			Dictionary<Tile, Tile> predecessors = new Dictionary<Tile, Tile>();
+			HashSet<Tile> visited = new HashSet<Tile>();
+
+			dist[start] = 0;
+
+			bool updateShortestDistance(Tile tile, int distance) {
+				if (!dist.ContainsKey(tile)) {
+					dist[tile] = distance;
+					return true;
+				} else if (dist[tile] > distance) {
+					dist[tile] = distance;
+					return true;
+				}
+				return false;
+			}
+
+			// TODO not O(n) search
+			KeyValuePair<Tile, int> getNextClosest() {
```

Comment:
See above about putting in a method.  Although it would take params, once both of these are moved, the top-level method would read easier - what follows this method essentially _is_ Dijkstra's, these are just the sub-components.

For now, I wouldn't worry about the O(n).  My textbook has descriptions of various options for these.  But I always anticipated that this would be an iterative process of making things more optimal.